docs/designs/data_transform.md

```diff
@@ -0,0 +1,41 @@
+
+Data transform contains two key parts: analyzer and transformer. Analyzer scans the entire data set and calculates the statistical values such as mean, min, variance and so on. Transformer combines the statistical value and the transform function to construct the concrete transform logic. And then it transforms the data records one by one.  
+[TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) is the open source solution for data transform in [TensorFlow Extended](https://www.tensorflow.org/tfx/guide). Users need write a python function preprocess_fn to define the preprocess logic. SQLFlow users prefer to write SQL instead of python. It's user unfriendly to SQLFlow users if we integrate TF Transform with SQLFlow directly.  
+From another point of view, SQL can naturally support statistical work just like analyzer. Feature column api can take charge of transform logic such as transformer. For dense column, we can use numeric_column and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use embedding_column to map the sparse value to embedding vector or use cross_column to do the feature crossing. We plan to use SQL and feature column together to do the data transform work.  
+
+Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode offline and in real time mode online. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+
+## Transform Expression in SQLFlow
+
+We can extend the SQLFlow syntax and add **TO TRANSFORM** keyword to describe the transform process. Let's take the following SQL expression for example. It trains a model to classify someone's income level using the [census income dataset](https://archive.ics.uci.edu/ml/datasets/Census+Income). The transform expression is **TO TRANSFORM STANDARDIZE(age) as age_std, NORMALIZE(capital_gain) as capital_gain_norm, BUCKETIZED(hours_per_week, bucket_num=10) as hours_per_week_bkt**. It will standardize the column *age* to the column *age_std*, normalize the column *capital_gain* to *capital_gain_norm*, bucketize the column *hours_per_week* to 10 buckets to the column *hours_per_week_bkt*. The output of transform will be passed to the **COLUMN** expression.  
+We add some built-in transform API and users can use them directly in the TRANSFORM expression. The Api set contains NORMALIZE, STANDARDIZE, BUCKETIZED, LOG and more to be added in the future.  
+
+```SQL
+SELECT *
+FROM census_income
+TO TRANSFORM STANDARDIZE(age) as age_std, NORMALIZE(capital_gain) as capital_gain_norm, BUCKETIZED(hours_per_week, bucket_num=10) as hours_per_week_bkt
```

Comment:
How to do feature crosses here?

Reply:
Why not put functions like `STANDARDIZE` in COLUMN clause? And what's the benifit of using `TO TRANSFORM`?

Reply:
> How to do feature crosses here?

For the feature cross, we will describe it in the COLUMN expression. The TRANSFORM expression only describes the transform logic for one single column specially needs statistical operation.

Reply:
I have the same question as @typhoonzero. Instead of adding `TO TRANSFORM`, why not
```
COLUMN NUMERIC(STANDARDIZE(age)), ...
```

Reply:
As we know, SQLFlow expression will be translated to a Couler WorkFlow. 
The **TO TRAIN xxx COLUMN xxx** expression will be converted to a single model training step in the WorkFlow.
I think TRANSFORM is a separate step from Training step. It analyzes the full data set to calculate the statistical value. And then it generate the concrete the feature column definition for data transformation. So I choose to put TRANSFORM in a separate expression.
What's more, generate the feature column from the TRANSFORM expression is the V1 solution. 
In the next step, I'm thinking we can transform the raw data and then write the transformed data into a temporary table, and then the training job will run based on the transformed data. In this way, transform should also be run separately before the training step.
So I choose to put TRANSFORM into a single expression.

Reply:
Hi, @Echo9573 could you please share some suggestion about the transform expression from the point view of data analyst. Which expression pattern is more user friendly? Using a separate TRANSFORM expression or add the transform API call in the COLUMN expression?

Reply:
After discussion, we use COLUMN NUMERIC(STANDARDIZE(age)) instead of adding a separate TO TRANSFORM expression.